plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs

```diff
@@ -131,68 +134,146 @@ pattern DefaultUniArray uniA =
-            DefaultUniByteString <- Just a2
-            Just Refl
-        geqStep DefaultUniString a2 = do
-            DefaultUniString <- Just a2
-            Just Refl
-        geqStep DefaultUniUnit a2 = do
-            DefaultUniUnit <- Just a2
-            Just Refl
-        geqStep DefaultUniBool a2 = do
-            DefaultUniBool <- Just a2
-            Just Refl
-        geqStep DefaultUniProtoList a2 = do
-            DefaultUniProtoList <- Just a2
-            Just Refl
-        geqStep DefaultUniProtoArray a2 = do
-            DefaultUniProtoArray <- Just a2
-            Just Refl
-        geqStep DefaultUniProtoPair a2 = do
-            DefaultUniProtoPair <- Just a2
-            Just Refl
-        geqStep (DefaultUniApply f1 x1) a2 = do
-            DefaultUniApply f2 x2 <- Just a2
-            Refl <- geqRec f1 f2
-            Refl <- geqRec x1 x2
-            Just Refl
-        geqStep DefaultUniData a2 = do
-            DefaultUniData <- Just a2
-            Just Refl
-        geqStep DefaultUniBLS12_381_G1_Element a2 = do
-            DefaultUniBLS12_381_G1_Element <- Just a2
-            Just Refl
-        geqStep DefaultUniBLS12_381_G2_Element a2 = do
-            DefaultUniBLS12_381_G2_Element <- Just a2
-            Just Refl
-        geqStep DefaultUniBLS12_381_MlResult a2 = do
-            DefaultUniBLS12_381_MlResult <- Just a2
-            Just Refl
-        geqStep DefaultUniValue a2 = do
-            DefaultUniValue <- Just a2
-            Just Refl
-        {-# INLINE geqStep #-}
-
-        geqRec :: DefaultUni a1 -> DefaultUni a2 -> Maybe (a1 :~: a2)
-        geqRec = geqStep
-        {-# OPAQUE geqRec #-}
+    geq = goStep where
+        -- Even though performance of this definition isn't particularly important, we still define
+        -- 'geq' manually instead of using 'deriveGEq', because the latter not only requires a
+        -- dependency on the 'dependent-sum-template' library that caused multiple problems, it also
+        -- creates a single recursive definition and having two instead is faster while not being
+        -- very annoying to implement and maintain.
+        --
+        -- Splitting the definition in two allows GHC to inline the initial step that appears
+        -- non-recursive to GHC, because recursion is hidden in the other function that is marked as
+        -- @NOINLINE@ and is chosen by GHC as a loop-breaker, see
+        -- https://wiki.haskell.org/Inlining_and_Specialisation#What_is_a_loop-breaker (we're not
+        -- really sure if this is a reliable solution, but if it stops working, we won't miss very
+        -- much and we've failed to settle on any other approach).
+        --
+        -- We use @NOINLINE@ instead of @OPAQUE@, because we don't actually care about the recursive
+        -- definition not being inlined, we just want it to be chosen as a loop breaker.
+        goStep, goRec :: DefaultUni a1 -> DefaultUni a2 -> Maybe (a1 :~: a2)
```

Comment:
Is there actual meaningful change here, or just renaming?

Reply:
Just renaming, I was playing with other stuff and decided that there should be a convention to call such functions `goStep` and `goRec`.